### PR TITLE
[core] Async submitting actor registerring

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

This PR introduced a separate state in actor creator to recording the actors being registering.
With this, we'll be able to register actors in an async way and also make certain calls waiting on the registration.

- ~~This PR also add a bunch of mock method for testing~~

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
@rkooo567 @ericl Could you verify the implementation from the high level looks good?

I'll add unit test cases once it's in the right direction.

File: src/ray/core_worker/transport/dependency_resolver.cc
Comment (ericl) [inline src/ray/core_worker/transport/dependency_resolver.cc:91]: @stephanie-wang I assume this captures all nested refs too?

File: src/ray/core_worker/transport/dependency_resolver.h
Comment (ericl) [inline src/ray/core_worker/transport/dependency_resolver.h:55]: Why not keep shared_ptr? Ray references have caused many destruction order issues.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (ericl) [inline src/ray/core_worker/transport/direct_task_transport.cc:117]: Can we feature flag the asynchrony?

File: src/ray/protobuf/common.proto
Comment (ericl) [inline src/ray/protobuf/common.proto:134]:
```suggestion
  DEPENDENCY_RESOLUTION_FAILED = 8;
```

Please spell check.

File: src/ray/core_worker/transport/dependency_resolver.cc
Comment (ericl) [inline src/ray/core_worker/transport/dependency_resolver.cc:91]: Is this actually capturing nested refs? For example, suppose

```
actor = Actor.remote()
nested_ref = ray.put(actor)
f.remote(nested_ref)
```

cc @stephanie-wang

**Review by ericl — COMMENTED** [review]
Overall seems reasonable, but some questions on nested ref handling.

**Comment by jovany-wang** [comment]
Hi @iycheng Does this PR aims to address https://github.com/ray-project/ray/issues/13900 ?

File: src/ray/core_worker/transport/dependency_resolver.h
Comment (iycheng) [reply src/ray/core_worker/transport/dependency_resolver.h:55]: I believe it gives a better understanding of ownership. Using T* or T& means that this class doesn't own this object. 

Destruction ordering issues should just be fixed. shared_ptr should only be used when sharing data across threads.

Some reference [here](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f7-for-general-use-take-t-or-t-arguments-rather-than-smart-pointers) and [here](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-unique)

**Comment by iycheng** [comment]
> Hi @iycheng Does this PR aims to address #13900 ?

@jovany-wang I think that issue has been fixed? cc @rkooo567 

For this PR, actor launching performance is slow because of sync registration, and this PR tried to fix this by converting it back to async way but still keep the function working.

File: src/ray/core_worker/transport/dependency_resolver.h
Comment (ericl) [reply src/ray/core_worker/transport/dependency_resolver.h:55]: How would you fix it for lambdas? I commonly see this pattern:

```
post([this]() {
   this->some_ref->something()
 });
```

This will generate a segfault if the callback is executed after the destruction of `this`. If the callback has a shared_ptr to this (shared_from_this()) or captures shared ptrs to `some_ref`, then this becomes much safer.

Comment (iycheng) [reply src/ray/core_worker/transport/dependency_resolver.h:55]: If `this` is shared across threads (like post to another thread), we can make this shared_from_this, but the member variable inside this doesn't need to be shared_ptr. It explicitly specifies the ownership.

If this is not shared across threads, maybe we should drain the task first?

File: src/ray/core_worker/transport/dependency_resolver.cc
Comment (iycheng) [reply src/ray/core_worker/transport/dependency_resolver.cc:91]: It won't capture the nested refs. Should I also add the support for this here?

Comment (ericl) [reply src/ray/core_worker/transport/dependency_resolver.cc:91]: Hmm, we should if it's easy. Alternatively we could add a barrier prior to returning from any ray.put() call? I think that would also solve it, without needing the complexity of nested id tracking.

**Comment by rkooo567** [comment]
I think that's not fixed. The original issue was that the cluster hangs due to https://github.com/ray-project/ray/pull/8045 (if the owner is dead before dependencies are resolved "the cluster hanged"). 

@iycheng I think as long as this test passes `test_actor_owner_worker_dies_before_dependency_ready`, this should be fine.

**Comment by rkooo567** [comment]
Btw, can you write a description how you handle this?

**Comment by iycheng** [comment]
> Btw, can you write a description how you handle this?

Added

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_task_transport.cc:112]:
```
This caused a problem: if the owner of the actor dies before dependencies
    // are resolved, the actor will never be created. But the actor handle may already be
    // passed to other workers
```

How do we handle this now? Are we not allowing to pass actor handles to other workers until registration is done?

File: src/ray/core_worker/transport/dependency_resolver.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/dependency_resolver.cc]: Sorry I don't have much context here, but is this to make sure that an actor is created before passing its handle to another task? I'm not sure if this step is necessary.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (iycheng) [reply src/ray/core_worker/transport/direct_task_transport.cc:112]: Yes, we are not passing it to other workers until it's done. It's considered as a dependence.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (iycheng) [reply src/ray/core_worker/transport/direct_actor_transport.cc:105]: There will be two errors: 
- actor task (for this one); 
- actor creation failure

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (iycheng) [reply src/ray/core_worker/transport/direct_task_transport.cc:117]: I can have a feature flag to turn asyn reg to sync reg. But I'd prefer going the async code path by default and turn it off in case of failures

**Comment by ericl** [comment]
Sounds good (feature flag on by default).

**Review by rkooo567 — COMMENTED** [review]
The approach makes sense to me. When we don't allow to pass handles to other workers until it is registered, are we just blocking the main thread?

File: src/ray/core_worker/transport/dependency_resolver.cc
Comment (iycheng) [reply src/ray/core_worker/transport/dependency_resolver.cc]: @stephanie-wang  sorry for miss your message here. This is to ensure the actor is **registered** before passing its handle to another task.

**Comment by iycheng** [comment]
I'm going to rebase once this merged https://github.com/ray-project/ray/pull/18265

File: src/ray/gcs/gcs_client/service_based_accessor.cc
Comment (iycheng) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:542]: this is for mock testing. somehow boost::optional won't work due to print issues.

File: src/ray/core_worker/transport/dependency_resolver.cc
Comment (ericl) [inline src/ray/core_worker/transport/dependency_resolver.cc:143]: Is this thread-safe?

Comment (ericl) [reply src/ray/core_worker/transport/dependency_resolver.cc:143]: (do we need mutexLock(mu_))?

File: src/ray/core_worker/core_worker.h
Comment (ericl) [inline src/ray/core_worker/core_worker.h:1329]: Comment.

**Review by ericl — COMMENTED** [review]
LGTM at high level, question on thread safety in the dependency resolver

**Comment by ericl** [comment]
Some failing tests.

File: src/ray/core_worker/transport/dependency_resolver.cc
Comment (iycheng) [reply src/ray/core_worker/transport/dependency_resolver.cc:143]: @ericl I think you are asking a really good question and I found out that I don't have a clear answer unless I'm checking the code e2e.

I think the issue here is the poorly defined threading model. I probably will re-investigate the fiber thing mentioned previously after this.

For this question:
CoreWorker::SubmitActorTask -> direct_actor_submitter_->SubmitTask(io pool) -> ResolveDependency 
CoreWorker::SubmitTask -> direct_task_submitter_->SubmitTask(io pool) -> ResolveDependency
CoreWorker::CreateActor -> direct_task_submitter_->SubmitTask -> ResolveDependency

So create actors doesn't work. I'll update create actor to post tasks into io pool like SubmitActorTask/SubmitTask is doing.

Comment (ericl) [reply src/ray/core_worker/transport/dependency_resolver.cc:143]: Ok. If we're relying on tasks running in a particular thread for safety, we could add asserts that the current thread id is a particular one, this would also help the readability of the code.

Comment (iycheng) [reply src/ray/core_worker/transport/dependency_resolver.cc:143]: I think that will help identify whether things are going wrong, but not very useful for the organization of the code and lead to poor readability. 

We should have a clear boundary between modules running in different threads and adopt a way that prevents developer from doing things incorrectly.

**Comment by iycheng** [comment]
stress_test_dead_actors

```
  success = 1
  total_time = 120.58714270591736
  avg_iteration_time = 1.20586816072464
  max_iteration_time = 3.2681970596313477
  min_iteration_time = 0.0715939998626709
  _runtime = 611.8985605239868
  _session_url = https://beta.anyscale.com/o/anyscale-internal/projects/prj_LFKjNprpiYt4AjAf1NdLDJwn/clusters/ses_p2v7zVyANTa9kR1SQ26Hf9TD
  _commit_url = https://ray-wheels.s3.us-west-2.amazonaws.com/test_wheels/async-actor/bab5bd877250da431bebd8745dd66085a86432bf/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
  _stable = True
```

many_actors
```
  actors_per_second = 427.87014721114275
  num_actors = 10000
  time = 23.37157678604126
  success = 1
  _runtime = 37.6427104473114
  _session_url = https://beta.anyscale.com/o/anyscale-internal/projects/prj_LFKjNprpiYt4AjAf1NdLDJwn/clusters/ses_MThcuD15AVSyWSnwkRn6n7Dh
  _commit_url = https://ray-wheels.s3.us-west-2.amazonaws.com/test_wheels/async-actor/bab5bd877250da431bebd8745dd66085a86432bf/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
  _stable = True
```

**Comment by iycheng** [comment]
Some benchmark here
now:
```
  actor_launch_time = 5.364205384999877
  actor_ready_time = 83.60731100200019
  total_time = 88.97151638700007
  num_actors = 10000
  success = 1
  _runtime = 102.01261568069458
  _session_url = https://beta.anyscale.com/o/anyscale-internal/projects/prj_LFKjNprpiYt4AjAf1NdLDJwn/clusters/ses_NWZTExNmQB3SNgshMP6H1Uzk
  _commit_url = https://ray-wheels.s3.us-west-2.amazonaws.com/test_wheels/async-actor/bab5bd877250da431bebd8745dd66085a86432bf/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
  _stable = True
```

vs

before:
```
  actor_launch_time = 179.710744493
  actor_ready_time = 14.067243351000116
  total_time = 193.7779878440001
  num_actors = 10000
  success = 1
  _runtime = 206.03042674064636
  _session_url = https://beta.anyscale.com/o/anyscale-internal/projects/prj_LFKjNprpiYt4AjAf1NdLDJwn/clusters/ses_5ccZB2Z97L7eiJiqNd2idYBj
  _commit_url = https://s3-us-west-2.amazonaws.com/ray-wheels/master/dee12be25386415a6b5a858fdb605a3457e90f66/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
  _stable = True
```

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
OK to merge once we verify the thread-safety of the dep resolution code.

**Comment by iycheng** [comment]
> OK to merge once we verify the thread-safety of the dep resolution code.

@ericl  I've added a wrapper to make sure variables are only accessed in the same thread.

File: src/ray/util/util.h
Comment (ericl) [inline src/ray/util/util.h:258]: How about calling this `ThreadPrivate` instead?

File: src/ray/util/util.h
Comment (ericl) [inline src/ray/util/util.h:292]:
```suggestion
        << "A variable private to thread XX was accessed in thread YY."
```

File: src/ray/util/util.h
Comment (ericl) [inline src/ray/util/util.h:287]:
```suggestion
      RAY_LOG(DEBUG) << "Created in thread: " << thread_name_;
```

**Review by ericl — CHANGES_REQUESTED** [review]

File: src/ray/util/util.h
Comment (iycheng) [reply src/ray/util/util.h:287]: It's actually not creating, it's the first time it's used. How about 'Accessed in thread: ..'

Comment (ericl) [reply src/ray/util/util.h:287]: Oh, that's a bit weird. Why not first time created?

File: src/ray/util/util.h
Comment (ericl) [inline src/ray/util/util.h:261]: Why not set the thread id in the constructor?

File: src/ray/util/util.h
Comment (ericl) [inline src/ray/util/util.h:292]:
```suggestion
        << "A variable private to thread " << thread_name_ << " was accessed in thread "
```

**Review by ericl — CHANGES_REQUESTED** [review]

File: src/ray/util/util.h
Comment (iycheng) [reply src/ray/util/util.h:261]: We are not monitoring where it's constructed, but where is used. They can be constructed in the main thread and used in io_worker thread.

**Review by ericl — APPROVED** [review]

**Pull request merged**
